Objects as API curated example

papafe opened the pull request:
This is only a draft, and it's here mostly to get initial feedback and to run some discussion about a couple of rough points I've met. 

There have been 2 main approaches we've talked about when working on this:
1) Generic request. There is only one request type, with two `RealmValue` properties representing the payload and the response. Those need to be realm object.
2) Specific request type. There are multiple request object types, as many as the "api calls". In this case the payload and response can be embedded objects

So far I've gone with nr 1, as we discussed. Overall, I think both have some cons:

Generic request:
PRO:
- Mostly makes sense, it's easy to see similarity with REST.
- Need one trigger for all requests, as there is only one request type.
- Payload and response are separate objects, so the trigger on the request trigger doesn't get called when these are being modified. This is nice for example when the payload is bound the UI and so it's continuously being modified but the request shouldn't be handled until later.
- One request type, easier to show in the UI (but this is because we're showing them at all in the UI in this example)
- It's possible to create interfaces to represent the payload and response types.
- We need one trigger for all requests

CONS:
- It's a little awkward to have payload and response to be realm objects instead of embedded objects, as they don't have meaning outside of a request.
- Need to subscribe to all possible payload and responses objects (double the amount).
- **Payload and response are separate objects, so it can happen that when the trigger runs, the request that contains them has the dbref for both is set, but they're not yet uploaded. We can't go around this easily.
- **The response is create in the trigger, and it has happened sometimes that the client app receives an "empty" response, without the properties set. This causes a crash when trying to get the value of an enum property that is wrapped around a string. The string is empty at this moment, so we get a parse exception.

Specific request
PRO
- Less flexible sync subscriptions (one per api type).
- The payload and response can be embedded objects.
- A little bit easier to work with the trigger, no need to create the response, add it to the collection, and create the dbref to put in the request as with the generic request approach.
- When the trigger is run, we can be sure that the payload is there (it's embedded)
- There shouldn't be problems with "empty" responses, because it's embedded. 

CONS:
- More difficult to show all request types in the UI. Admittedly, this is an issue only for this app
- The payload is embedded, so the backend trigger gets called all the time while the user is creating an order (due to UI bindings), for example.
- It's very awkward to create an interface for the request type, because payload and response are not realm values
- We need to one trigger per API call.

So... overall there are issues with both. As I said, I went with the generic request approach, but it's still possible to switch the other if we decide it's a better fit.
Overall I think that the biggest issues for me are the ones marked with **. And I will leave comments in the code so we can talk about those issues more specifically

papafe committed:
Objects as API stub

papafe committed:
Added services, login page and busy popup

papafe committed:
Added stub for pages

papafe committed:
Added stub for views

papafe committed:
Various fixes

papafe committed:
Added creator id

papafe committed:
Various fixes

papafe committed:
Fixed UI

papafe committed:
Added mapto

papafe committed:
Removed extra parameter and simplified names

papafe committed:
Finished renaming

papafe committed:
Modifications

papafe committed:
Various fixes

papafe committed:
Removed tabs

papafe committed:
Corrected UI

papafe committed:
Added stub for cancel order

papafe committed:
Removed unnecessary

papafe committed:
Added multiple things

papafe committed:
Removed id

papafe commented on ObjectsAsAPI/Models/CreateOrder.cs:
```diff
@@ -0,0 +1,54 @@
+
+    public ResponseStatus Status
+    {
+        get => string.IsNullOrEmpty(_Status) ? ResponseStatus.Approved : Enum.Parse<ResponseStatus>(_Status);
```
This here has been done like to avoid the situation in which we get a response, and the properties are not set yet. If `_Status` is empty, then we'll get an exception here.
Responses are created on the backend in a trigger, and I suppose that the "create response" and "fill response" changes can be split into different changesets.

papafe commented on ObjectsAsAPI/ViewModels/OrderViewModel.cs:
```diff
@@ -0,0 +1,54 @@
+    [RelayCommand]
+    public async Task CancelOrder()
+    {
+        var realm = RealmService.GetMainThreadRealm();
```
In this method we first set the request to draft and only later we set it to pending, so that we can be sure that when the trigger is called (when `Status == Pending`), the payload has been uploaded already.
In a previous version we were creating the request directly after the dialog, with status set to pending. Sometimes this would cause issues on the backend, because the trigger couldn't find the payload yet.

papafe commented:
I've had a discussion with @nirinchev and we agreed to move to using specific classes for requests, as this should solve most of the issues that are present now

papafe committed:
Removed some comments

papafe committed:
Stub for new style

papafe committed:
Modified other request

papafe committed:
Various correction (grouped)

papafe committed:
Various corrections

papafe committed:
Various small fixes

papafe committed:
Added Atlas folder

papafe committed:
Added Readme + rename

papafe commented:
@nirinchev This is ready for review. I tried to keep the README short, didn't want to make it too long.

nirinchev commented on ObjectsAsAPI/AppShell.xaml.cs:
```diff
@@ -0,0 +1,16 @@
+public partial class AppShell : Shell
+{
+	public AppShell()
+	{
```
Mixed tabs and spaces

nirinchev commented on ObjectsAsAPI/AtlasApp/functions/CancelOrderRequestHandler.js:
```diff
@@ -0,0 +1,62 @@
+      response = {
+        "status": "Rejected",
+        "orderId": orderId,
+        "rejectedReason": "There too few items in this order to be cancelled!"
```
```suggestion
        "rejectedReason": "There are too few items in this order to be cancelled!"
```

nirinchev commented on ObjectsAsAPI/AtlasApp/functions/CancelOrderRequestHandler.js:
```diff
@@ -0,0 +1,62 @@
+      "$set": {
+        "status" : "Handled",
+        "response": response 
+      }
```
Haven't gone through everything yet, but just skimming it, seems like we have a few too many `status` fields here. Perhaps it's needed, but I wonder if we're not overdoing something. It's also awkward that they're all different values.

nirinchev commented on ObjectsAsAPI/AtlasApp/functions/CreateOrderRequestHandler.js:
```diff
@@ -0,0 +1,65 @@
+
+    var response;
+
+    if(payload.content.items == undefined) {
```
Can there be 0 items here?

papafe replied:
Mmh, maybe if we add an element and then remove it, I'll check

nirinchev commented on ObjectsAsAPI/Models/CancelOrderRequest.cs:
```diff
@@ -0,0 +1,122 @@
+{
+    [PrimaryKey]
+    [MapTo("_id")]
+    public ObjectId Id { get; set; } = ObjectId.GenerateNewId();
```
```suggestion
    public ObjectId Id { get; private set; } = ObjectId.GenerateNewId();
```

nirinchev commented on ObjectsAsAPI/Models/CancelOrderRequest.cs:
```diff
@@ -0,0 +1,122 @@
+    public string CreatorId { get; set; }
+
+    [MapTo("status")]
+    private string _Status { get; set; } = null!;
```
Should this be initialized to `null` or to some default value of `RequestStatus`?

papafe replied:
We could do it for safety

nirinchev commented on ObjectsAsAPI/Models/CancelOrderRequest.cs:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    [MapTo("createdAt")]
+    public DateTimeOffset CreatedAt { get; set; }
```
Should this have a setter?

papafe replied:
Probably not.

nirinchev commented on ObjectsAsAPI/Models/CancelOrderRequest.cs:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    // Used in the UI
+    public string? StatusString
+    {
+        get
+        {
+            return Status switch
+            {
+                RequestStatus.Draft => "Draft",
+                RequestStatus.Pending => "Pending",
+                RequestStatus.Handled => $"{Response?.Status}{(string.IsNullOrEmpty(Response?.RejectedReason) ? "" : $": \"{Response?.RejectedReason}\"")}",
+                _ => throw new NotImplementedException(),
+            };
+        }
+    }
```
I don't want to be a MVVM purist, but shouldn't this and `Description` be handled by a value converter? Also, wouldn't we run into issues where we won't raise property change notifications correctly here?

papafe replied:
Yes, it could be handled by a value converter, I wanted to make this simpler because we need a `MultiValueConveter` for these I think. It looks kinda ugly as it is now though. 

Regarding the notifications, that's what I'm trying to do with `OnPropertyChanged` down here

nirinchev replied:
While we don't need to design this in a way that's super robust and future-proof, the current implementation of `OnPropertyChanged` is somewhat heavily tied to the business logic, where we make the assumption that `Response` only changes when `Status` changes. Technically, it's possible that `Response` or `Response.RejectedReason` change independently of `Status` in which case `StatusString` will not emit notifications. That being said, if a multivalueconverter further complicates things, I'd be fine with leaving the implementation as it is and just adding a comment in `OnPropertyChanged` for why checking for `Status` is sufficient.

papafe replied:
Agree with that

nirinchev commented on ObjectsAsAPI/Models/CancelOrderRequest.cs:
```diff
@@ -0,0 +1,122 @@
+    }
+}
+
+public partial class CancelOrderPayload : IEmbeddedObject, IPayload
+{
+    [MapTo("orderId")]
+    public ObjectId OrderId { get; set; }
+}
+
+public partial class CancelOrderResponse : IEmbeddedObject, IResponse
+{
+    [MapTo("orderId")]
+    public ObjectId OrderId { get; set; }
+
+    [MapTo("status")]
+    private string _Status { get; set; } = null!;
+
+    public ResponseStatus Status
+    {
+        get => Enum.Parse<ResponseStatus>(_Status);
+        private set => _Status = value.ToString();
+    }
+
+    [MapTo("rejectedReason")]
+    public string? RejectedReason { get; private set; }
+}
```
Since we now have dedicated classes, do we need the payload and response objects? Seems like a bit of an overkill and data duplication.

papafe replied:
Yes, we could probably simplify things if we remove the embedded objects and just put everything on the request object. I like this approach more though, as we have a clearer separation between the different parts. 

I'm not sure about data duplication here though. What were you thinking about?

nirinchev replied:
I'm fine with having the embedded objects, as long as there's meaningful information to store in them. Re: data duplication - I can see the `OrderId` exists in both the request payload and the response and I feel the `Status` on the response could be merged with the status on the request.

papafe replied:
Agree with those points. The `OrderId` in the response here is a leftover from the previous implementation, I'll remove it

nirinchev commented on ObjectsAsAPI/README.md:
```diff
@@ -0,0 +1,266 @@
+
+**ObjectsAsAPI** is a sample MAUI application that shows how to it is possible to use the insertion or update of Realm objects as a way to simulate a backend API call once the objects are synchronized on Atlas. In practice, this example shows how is it possible to create specific Realm objects representing API calls (**objects-as-api**), and how it is possible to use [Database Triggers]((https:// ...
+
+For this example application, let us imagine the use case of a repairman that works on ships, where it cannot be guaranteed a stable internet connection. The repairman needs to create orders of all the pieces he needs for his repairs as he goes through the ship. In a later section we will take a look at how the app is structured to handle API calls all within Realm and Atlas Device Sync.
```
```suggestion
For this example application, let us imagine the use case of a technician who works on ships, where a stable internet connection cannot be guaranteed. They need to create orders for all the pieces they need for the repairs as they go through the ship. In a later section we will take a look at how the app is structured to handle API calls all within Realm and Atlas Device Sync.
```

nirinchev commented on ObjectsAsAPI/README.md:
```diff
@@ -0,0 +1,266 @@
+    <img src="Images/app.jpeg">
+</p>
+
+The application allows the user to create orders, containing an arbitrary number of items, and possibly cancel them. The main screen of the app (shown in the screenshot) contains the list of the orders on the top part of the screen, while at the bottom it shows the *API Requests* that are used to create or cancel order. The page contains also a button that allows to stop and resume synchronizatio ...
```
```suggestion
The application allows the user to create orders, containing an arbitrary number of items, and possibly cancel them. The main screen of the app (shown in the screenshot) contains the list of the orders on the top part of the screen, while at the bottom it shows the *API Requests* that are used to create or cancel an order. The page contains also a button that allows you to stop and resume synchronization, to simulate the lack of internet connection.
```

nirinchev commented on ObjectsAsAPI/README.md:
```diff
@@ -0,0 +1,266 @@
+
+The application allows the user to create orders, containing an arbitrary number of items, and possibly cancel them. The main screen of the app (shown in the screenshot) contains the list of the orders on the top part of the screen, while at the bottom it shows the *API Requests* that are used to create or cancel order. The page contains also a button that allows to stop and resume synchronizatio ...
+
+Clicking on the "+" button allows to create a new order, while clicking on an order will show its state, and allows to cancel it. Clicking on a request, instead, allows to see its state, and eventually cancel the request itself. For simplicity, it's also possible to swipe on all objects and delete them.
```
```suggestion
Clicking on the "+" button allows you to create a new order, while clicking on an order will show its state, and allows to cancel it. Clicking on a request, instead, allows to see its state, and eventually cancel the request itself. For simplicity, it's also possible to swipe on all objects and delete them.
```

nirinchev commented on ObjectsAsAPI/README.md:
```diff
@@ -0,0 +1,266 @@
+## The API
+
+In this application we want to show how the synchronization of Realm objects can be treated as an API call when combined with Database Triggers.
+For this, we have created `Request` objects, that represents the equivalent of an API call. 
```
```suggestion
For this, we have created `Request` objects, that represent the equivalent of an API call. 
```

nirinchev commented on ObjectsAsAPI/README.md:
````diff
@@ -0,0 +1,266 @@
+}
+```
+
+There a couple of things to notice here:
````
```suggestion
There are a couple of things to notice here:
```

nirinchev commented on ObjectsAsAPI/README.md:
```diff
@@ -0,0 +1,266 @@
+{
+    Approved,
+    Rejected,
+}
```
Don't have super strong opinions here, but we can probably simplify the design a little bit by combining these statuses - we can have a single request status that includes `Approved` and `Rejected` instead of `Handled` and move that info outside of the `IResponse` interface.

papafe replied:
Yes, probably it would simplify things.

nirinchev reviewed (commented):
Looks quite reasonable for the most part. A few minor comments from me and a suggestion to move things around the models a little bit to make them more straightforward to understand on a first pass.

papafe committed:
Apply suggestions from code review

Co-authored-by: Nikola Irinchev <irinchev@me.com>

papafe committed:
Update ObjectsAsAPI/README.md

Co-authored-by: Nikola Irinchev <irinchev@me.com>

papafe committed:
Small fixes to model and spaces

papafe committed:
Various fixes

papafe committed:
Simplification

papafe committed:
Moved status to value converter

papafe committed:
Various improvements

papafe committed:
Fixed naming

papafe committed:
Fixed backend

papafe committed:
Fixed function and readme

papafe commented:
@nirinchev I think I've addressed everything:
- `Status` now it on the request, so it's simple
- Removed the embedded payload and response, so it's simplified furtherly
- Added the converter for `Status` and `Description`, so we don't pollute the models
- Fixed the Readme file to reflect those changes
- Added a line for this project in the Readme of the repo

papafe committed:
Merge branch 'main' into fp/object-as-api

* main:
  Add Atlas Search demo (#5)
  Added error handling example (#6)

papafe committed:
Added line to main readme

nirinchev commented on ObjectsAsAPI/AtlasApp/functions/CreateOrderRequestHandler.js:
```diff
@@ -0,0 +1,62 @@
+      
+      const orderId = (await orderCollection.insertOne(order)).insertedId;
+      
+      orderRef = {
+        $ref: orderCollectionName,
+        $id: orderId,
+        $db: databaseName,
+      };
```
Does this need to be a dbref?

papafe replied:
Is there another way of doing this? In the docs I can't find anything particularly useful so far

nirinchev replied:
I think just setting the id is enough because this is not a mixed field, so the schema enforces that it links to the order collection.

papafe replied:
You were right, corrected it.

nirinchev commented on ObjectsAsAPI/Models/CancelOrderRequest.cs:
```diff
@@ -0,0 +1,52 @@
+
+    public CancelOrderRequest()
+    {
+        if (RealmService.CurrentUser == null)
+        {
+            throw new Exception("Login before using models!");
+        }
```
Would it make sense to instead pass in the user in the ctor?

papafe replied:
Mmh, I think it's cleaner like this, and it's easier to use the objects. Once they've been instantiated both the user and the creation date are already set

nirinchev reviewed (approved):
Looks good - two minor nits from me.

papafe committed:
Corrected function
